### PR TITLE
Iterator comparison and memory bug fixes

### PR DESCRIPTION
**Pull request by reuk**
I found a load of iterator comparison issues in CBMC, which this PR fixes. Includes the fix from #1610, which I'm closing because the issues fixed by this PR are all so similar.

### CODE REVIEW

File: src/analyses/dependence_graph.cpp
Comment (smowton) [inline src/analyses/dependence_graph.cpp]: Suggest commenting that just comparing function is not sufficient because of recursive calls

**Review by smowton — APPROVED** [review]
Suggest adding a comment to forestall the obvious but incorrect future "improvement"; otherwise lgtm

**Comment by reuk** [comment]
I'm running ASAN and UBSAN too. Seeing as all these fixes are pretty small, I'll carry on adding fixes here until I run out of issues or the PR reaches ~200 lines, whichever comes ~last~ first

**Review by smowton — APPROVED** [review]

File: src/analyses/ai.h
Comment (tautschnig) [inline src/analyses/ai.h]: I'd suggest that this be moved to goto-programs/goto_program_template.h, which already has some related comparison code.

File: src/cpp/cpp_typecheck_compound_type.cpp
Comment (tautschnig) [inline src/cpp/cpp_typecheck_compound_type.cpp]: You might use `value.has_operands()`.

File: src/goto-instrument/cover.cpp
Comment (tautschnig) [inline src/goto-instrument/cover.cpp]: I believe this should be expressed via `CHECK_RESULT` or `PRECONDITION` as empty decisions would seem wrong at this point.

File: src/goto-programs/graphml_witness.cpp
Comment (tautschnig) [inline src/goto-programs/graphml_witness.cpp]: In line with the earlier comment: equality over iterators is generally useful, thus use the to-be-introduced function.

**Comment by tautschnig** [comment]
I've restarted the Travis job, but I'm concerned that we might be running into timeouts on a regular basis with those additional changes. (Those changes aren't necessarily the one root cause, they just contribute to it.)

File: src/goto-instrument/cover.cpp
Comment (reuk) [reply src/goto-instrument/cover.cpp]: I suggest this is fixed by someone with more domain knowledge. I made this change because decisions *isn't* always populated, but finding out why it's sometimes empty goes beyond the scope of a simple memory-bug fix. Changing this to an INVARIANT will cause some of the tests to fail, which in turn will prevent this PR being merged.

**Comment by tautschnig** [comment]
It seems further bugs remain: Travis reports "Error: attempt to dereference a singular iterator." for all the tests doing full slicing.

**Comment by reuk** [comment]
Yes, I'm still working on this. I think that's the only bug left though (at least of the issues uncovered by the regression and unit tests).

File: src/goto-instrument/cover.cpp
Comment (tautschnig) [reply src/goto-instrument/cover.cpp]: I'd suggest the following:
1. Use `git blame` to identify the author of those lines.
2. Use (some variation of) `INVARIANT`.
3. Mark all failing tests as `KNOWNBUG`.
4. File an issue and assign it to the author identified in 1.

File: src/goto-programs/goto_program_template.h
Comment (tautschnig) [inline src/goto-programs/goto_program_template.h]: Why is this a better hash function, or rather: what bug is being "fixed" here?

Comment (reuk) [reply src/goto-programs/goto_program_template.h]: It appears that instruction location numbers were being modified while holding iterators to those same instructions in an unordered map. This breaks the map invariants.

Comment (reuk) [reply src/goto-programs/goto_program_template.h]: To be clear: this is the cause of the remaining bug in the slicer

File: src/goto-programs/graphml_witness.cpp
Comment (tautschnig) [inline src/goto-programs/graphml_witness.cpp]: Needs a `NOLINT`

File: src/goto-programs/goto_program_template.h
Comment (tautschnig) [reply src/goto-programs/goto_program_template.h]: Oh, very nice catch (and likely a bug that I had introduced, sorry...). Would you mind giving a somewhat more detailed explanation in the commit message?

**Comment by reuk** [comment]
I think this is ready for re-review. I've created #1622, which describes the error with dereferencing a singular `begin()` iterator in cover.cpp.

**Review by tautschnig — APPROVED** [review]

**Comment by tautschnig** [comment]
Waiting for CI to complete before doing the merge.

**Comment by tautschnig** [comment]
@chrisr-diffblue @martin-cs @peterschrammel @forejtv @thk123 @romainbrenguier please confirm that @smowton and I have been right to approve this.

**Comment by reuk** [comment]
I'd really like to get this merged today, or to continue working on any feedback. Pinging @chrisr-diffblue, @martin-cs, @peterschrammel, @forejtv, @thk123, @romainbrenguier.

**Review by chrisr-diffblue — APPROVED** [review]
It would be nice if the comments in the KNOWNBUG tests referenced #1622, but I'm be happy for this to go in as is.

**Comment by reuk** [comment]
@tautschnig I've added a reference to #1622 in the test descriptions. That's all that's changed since the last successful CI run. Would you merge please?

**Comment by tautschnig** [comment]
It seems we should still get more input from code owners: @forejtv, @peterschrammel, @thk123, @martin-cs, @romainbrenguier.

**Review by romainbrenguier — APPROVED** [review]

**Comment by tautschnig** [comment]
Which leaves us with @forejtv, @peterschrammel, @thk123 to contribute a review...

**Comment by tautschnig** [comment]
... which I'm now daring to override.

**Comment by martin-cs** [comment]
I'm late to the party on this one, so I guess the onus is on me to fix this BUT:

The changes to the ai_base_domaint interface are not an appropriate solution and should be reverted.

I'm happy to look at the original problem and try to find alternative solution but we can't leave this.

**Comment by reuk** [comment]
You've said that the interface is problematic, but haven't explained why. Could you expand on that a bit please?

I discussed this change with @smowton, and it seems reasonable to me. There are a fixed number of contexts in which that method may be called, and the enum just distinguishes different cases. This is more robust than iterator comparisons, especially because in many cases the iterators don't belong to the same collection, leading to undefined behaviour. This change fixes actual bugs, and so I'd strongly advise against introducing a revert commit for the change. Indeed, such a change would no longer pass CI because of the additional checks introduced by this PR. However, an alternative solution that doesn't depend on iterator comparisons might be appropriate.

**Pull request merged**
